File: laravel_application/app/Models/Product.php
<?php

namespace App\Models;

use Illuminate\Database\Eloquent\Factories\HasFactory;
use Illuminate\Database\Eloquent\Model;
use Illuminate\Database\Eloquent\SoftDeletes;

class Product extends Model
{
    use HasFactory, SoftDeletes;

    protected $fillable = [
        'merchant_id',
        'category_id',
        'name',
        'name_ar',
        'description',
        'description_ar',
        'price',
        'original_price',
        'discount_percentage',
        'stock',
        'sku',
        'weight',
        'dimensions',
        'is_active',
        'is_featured',
        'views_count',
        'sales_count',
    ];

    protected $casts = [
        'price' => 'decimal:2',
        'original_price' => 'decimal:2',
        'discount_percentage' => 'integer',
        'stock' => 'integer',
        'weight' => 'decimal:2',
        'is_active' => 'boolean',
        'is_featured' => 'boolean',
        'views_count' => 'integer',
        'sales_count' => 'integer',
    ];

    // Relationships
    public function merchant()
    {
        return $this->belongsTo(Merchant::class);
    }

    public function category()
    {
        return $this->belongsTo(Category::class);
    }

    public function images()
    {
        return $this->hasMany(ProductImage::class);
    }

    public function variants()
    {
        return $this->hasMany(ProductVariant::class);
    }

    public function reviews()
    {
        return $this->hasMany(Review::class);
    }

    public function orderItems()
    {
        return $this->hasMany(OrderItem::class);
    }

    // Accessors
    public function getMainImageAttribute()
    {
        return $this->images()->where('is_primary', true)->first()?->image_url
            ?? $this->images()->first()?->image_url;
    }

    public function getAverageRatingAttribute()
    {
        return $this->reviews()->avg('rating') ?? 0;
    }

    public function getReviewsCountAttribute()
    {
        return $this->reviews()->count();
    }

    public function getFinalPriceAttribute()
    {
        if ($this->discount_percentage > 0) {
            return $this->price * (1 - $this->discount_percentage / 100);
        }
        return $this->price;
    }

    public function getIsOutOfStockAttribute()
    {
        return $this->stock <= 0;
    }

    public function getIsNewAttribute()
    {
        return $this->created_at->diffInDays(now()) <= 30;
    }

    // Methods
    public function decrementStock($quantity)
    {
        $this->decrement('stock', $quantity);
    }

    public function incrementStock($quantity)
    {
        $this->increment('stock', $quantity);
    }

    public function incrementViews()
    {
        $this->increment('views_count');
    }

    public function incrementSales()
    {
        $this->increment('sales_count');
    }

    // Scopes
    public function scopeActive($query)
    {
        return $query->where('is_active', true);
    }

    public function scopeFeatured($query)
    {
        return $query->where('is_featured', true);
    }

    public function scopeInStock($query)
    {
        return $query->where('stock', '>', 0);
    }

    public function scopeByCategory($query, $categoryId)
    {
        return $query->where('category_id', $categoryId);
    }

    public function scopePriceRange($query, $min, $max)
    {
        return $query->whereBetween('price', [$min, $max]);
    }

    public function scopeSearch($query, $search)
    {
        return $query->where(function($q) use ($search) {
            $q->where('name', 'like', "%{$search}%")
              ->orWhere('name_ar', 'like', "%{$search}%")
              ->orWhere('description', 'like', "%{$search}%")
              ->orWhere('description_ar', 'like', "%{$search}%");
        });
    }
}

File: flutter_app/lib/models/product.dart
class Product {
  final String id;
  final String name;
  final String nameAr;
  final String description;
  final String descriptionAr;
  final double price;
  final double originalPrice;
  final double discountPercentage;
  final int stock;
  final String? mainImage;
  final List<String> images;
  final double rating;
  final int reviewsCount;
  final String categoryId;
  final String merchantId;
  final bool isActive;
  final bool isFeatured;

  Product({
    required this.id,
    required this.name,
    required this.nameAr,
    required this.description,
    required this.descriptionAr,
    required this.price,
    this.originalPrice = 0,
    this.discountPercentage = 0,
    required this.stock,
    this.mainImage,
    this.images = const [],
    this.rating = 0,
    this.reviewsCount = 0,
    required this.categoryId,
    required this.merchantId,
    this.isActive = true,
    this.isFeatured = false,
  });

  factory Product.fromJson(Map<String, dynamic> json) {
    return Product(
      id: json['id'].toString(),
      name: json['name'] ?? '',
      nameAr: json['name_ar'] ?? '',
      description: json['description'] ?? '',
      descriptionAr: json['description_ar'] ?? '',
      price: double.tryParse(json['price'].toString()) ?? 0,
      originalPrice: double.tryParse(json['original_price']?.toString() ?? '0') ?? 0,
      discountPercentage: double.tryParse(json['discount_percentage']?.toString() ?? '0') ?? 0,
      stock: int.tryParse(json['stock']?.toString() ?? '0') ?? 0,
      mainImage: json['main_image'],
      images: json['images'] != null
          ? (json['images'] as List).map((img) => img['image_url'].toString()).toList()
          : [],
      rating: double.tryParse(json['rating']?.toString() ?? '0') ?? 0,
      reviewsCount: int.tryParse(json['reviews_count']?.toString() ?? '0') ?? 0,
      categoryId: json['category_id'].toString(),
      merchantId: json['merchant_id'].toString(),
      isActive: json['is_active'] ?? true,
      isFeatured: json['is_featured'] ?? false,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'name': name,
      'name_ar': nameAr,
      'description': description,
      'description_ar': descriptionAr,
      'price': price,
      'original_price': originalPrice,
      'discount_percentage': discountPercentage,
      'stock': stock,
      'main_image': mainImage,
      'images': images,
      'rating': rating,
      'reviews_count': reviewsCount,
      'category_id': categoryId,
      'merchant_id': merchantId,
      'is_active': isActive,
      'is_featured': isFeatured,
    };
  }
}
